xarray/core/dataarray.py

```diff
@@ -576,6 +576,33 @@ def reset_coords(self, names=None, drop=False, inplace=False):
             dataset[self.name] = self.variable
             return dataset
 
+    def __dask_graph__(self):
+        return self._variable.__dask_graph__()
```

Comment:
OK, that will be a bit tricky.  My guess is that it might be simpler to just account for all of the possible dask things explicitly, as we do in dataset.  Otherwise we're converting to and from datasets in each of the `__dask_foo__` methods, and I would not be surprised to run into oddness there.  I'm not sure though.

Can you recommend a test case that includes dask arrays in the coordinates?

Reply:
For some reason this fails when I use dask.arrays for coordinates

```python
    coord = da.arange(8, chunks=(4,))
    data = da.random.random((8, 8), chunks=(4, 4)) + 1
    array = DataArray(data,
                      coords={'x': coord, 'y': coord},
                      dims=['x', 'y'])
```

Replacing coords with a numpy array works fine.

Reply:
@mrocklin - what is the failure you are referring to?

Reply:
```python
In [1]: import xarray 

In [2]: import dask.array as da

In [3]:     coord = da.arange(8, chunks=(4,))
   ...:     data = da.random.random((8, 8), chunks=(4, 4)) + 1
   ...:     array = xarray.DataArray(data,
   ...:                       coords={'x': coord, 'y': coord},
   ...:                       dims=['x', 'y'])
   ...: 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-b90a33ebf436> in <module>()
      3 array = xarray.DataArray(data,
      4                   coords={'x': coord, 'y': coord},
----> 5                   dims=['x', 'y'])

/home/mrocklin/workspace/xarray/xarray/core/dataarray.py in __init__(self, data, coords, dims, name, attrs, encoding, fastpath)
    227 
    228             data = as_compatible_data(data)
--> 229             coords, dims = _infer_coords_and_dims(data.shape, coords, dims)
    230             variable = Variable(dims, data, attrs, encoding, fastpath=True)
    231 

/home/mrocklin/workspace/xarray/xarray/core/dataarray.py in _infer_coords_and_dims(shape, coords, dims)
     68     if utils.is_dict_like(coords):
     69         for k, v in coords.items():
---> 70             new_coords[k] = as_variable(v, name=k)
     71     elif coords is not None:
     72         for dim, coord in zip(dims, coords):

/home/mrocklin/workspace/xarray/xarray/core/variable.py in as_variable(obj, name)
     94                             '{}'.format(obj))
     95     elif utils.is_scalar(obj):
---> 96         obj = Variable([], obj)
     97     elif getattr(obj, 'name', None) is not None:
     98         obj = Variable(obj.name, obj)

/home/mrocklin/workspace/xarray/xarray/core/variable.py in __init__(self, dims, data, attrs, encoding, fastpath)
    275         """
    276         self._data = as_compatible_data(data, fastpath=fastpath)
--> 277         self._dims = self._parse_dimensions(dims)
    278         self._attrs = None
    279         self._encoding = None

/home/mrocklin/workspace/xarray/xarray/core/variable.py in _parse_dimensions(self, dims)
    439             raise ValueError('dimensions %s must have the same length as the '
    440                              'number of data dimensions, ndim=%s'
--> 441                              % (dims, self.ndim))
    442         return dims
    443 

ValueError: dimensions () must have the same length as the number of data dimensions, ndim=1
```

Reply:
My objective here is to produce a case where a data array has dask arrays in its coordinates so that I can write code to handle such cases.

Reply:
@mrocklin - something funny is going on here. I'm going to open a separate issue.

Reply:
In the short term, this may help you move forward:

```Python
In [21]: x = xr.Variable('x', da.arange(8, chunks=(4,)))
    ...: y = xr.Variable('y', da.arange(8, chunks=(4,)) * 2)
    ...: data = da.random.random((8, 8), chunks=(4, 4)) + 1
    ...: array = xr.DataArray(data,
    ...:                      dims=['x', 'y'])
    ...: array.coords['xx'] = x
    ...: array.coords['yy'] = y
    ...:
    ...:

In [22]: array
Out[22]:
<xarray.DataArray 'add-a034ba104341d3cca6b28ad7bf059b14' (x: 8, y: 8)>
dask.array<shape=(8, 8), dtype=float64, chunksize=(4, 4)>
Coordinates:
    xx       (x) int64 dask.array<shape=(8,), chunksize=(4,)>
    yy       (y) int64 dask.array<shape=(8,), chunksize=(4,)>
Dimensions without coordinates: x, y
```

Reply:
See #1684 for more information and #1685 for the fix.

Reply:
Thanks @jhamman , using your suggested code and @shoyer 's suggestion to depend on the DataSet implementation I think that this is now resolved.